lib/containers/export/ecl/list.hpp

```diff
@@ -0,0 +1,284 @@
+//! \file
+//! \brief The intrusive list interface
+//! \todo Examples
+#ifndef ECL_INTRUSIVE_LIST_
+#define ECL_INTRUSIVE_LIST_
+
+#include <cstddef>
+#include <type_traits>
+
+namespace ecl
+{
+
+// Give a star for this!
+// Taken from https://gist.github.com/graphitemaster/494f21190bb2c63c5516
+/// \todo move this to right module
+template< typename T1, typename T2 >
+inline typename std::enable_if< std::is_literal_type< T2 >::value, size_t >::type
+constexpr offset_of(T1 T2::*member) {
+    constexpr T2 object {};
+    return size_t(&(object.*member)) - size_t(&object);
+}
+
+/// \todo move this to right module
+template< typename T1, typename T2 >
+inline size_t constexpr offset_of(T1 T2::*member) {
+    return reinterpret_cast< size_t >(&(((T2*)0)->*member));
+}
+
+//! Intrusive, double-linked, circular list node.
+//! \details Any class can embed intrusive list node by composing it.
+class list_node
+{
+public:
+    //! Constructs emtpy list node.
+    list_node();
+    //! Destructs list node, unlinking itself from any list.
+    ~list_node();
+
+    //! Adds a node after this node.
+    //! \param[in] Node to add in the list
+    void add_after(list_node &node);
+
+    //! Adds a node before this node.
+    //! \param[in] Node to add in the list
+    void add_before(list_node &node);
+
+    //! Unlinks this node from a list.
+    //! \details Has no effect if list is empty.
+    void unlink();
+
+    //! Checks if node is connected to any other node.
+    bool linked() const;
+
+    //! Returns a pointer to a next list node.
+    list_node* next() const;
+
+    //! Returns a pointer to a previous list node.
+    list_node* prev() const;
+
+private:
+    list_node *m_next; //!< Next node in a list. Equals to this if the list is empty.
+    list_node *m_prev; //!< Previous node in a list. Equals to this if the list is empty.
+};
+
+
+//------------------------------------------------------------------------------
+
+
+list_node::list_node()
```

Comment:
Implementing the class in the header file is not ok unless these are inline methods (they are not).